### PR TITLE
21 check distribution of gevs

### PR DESCRIPTION
**Pull request by LiineKasak**

### CODE REVIEW

**Review by TaoDFang — APPROVED** [review]

**Comment by Sum02dean** [comment]
Neither of your suggestions was my intention. I want to compare all GEX values between the train and validation sets.

This means concatenating train data from X1 and X2 cell lines as one set, and concatenating the validation data from X1 and X2 as the other set.

The intention is to observe ( *all GEX values*) whether a significant shift in the distribution of GEX values exists. Not between chromosomes, or between cell lines, but across them.

Of course, we can also subset using the stratifications you suggest too...

An issue that commonly arises in ML is out of distribution (OOD) validation sets which result in some type of training bias.

It is important to see if the validation sets and training sets fall within the same distribution space.

**Comment by Sum02dean** [comment]
Neither of your suggestions was my intention. I want to compare all GEX values between the train and validation sets.

This means concatenating train data from X1 and X2 cell lines as one set, and concatenating the validation data from X1 and X2 as the other set.

The intention is to observe ( *all GEX values*) whether a significant shift in the distribution of GEX values exists. Not between chromosomes, or between cell lines, but across them.

Of course, we can also subset using the stratifications you suggest too...

An issue that commonly arises in ML is out of distribution (OOD) validation sets which result in some type of training bias.

It is important to see if the validation sets and training sets fall within the same distribution space.



To compare distributions across Chromosomes splits or cell-line splits. Simply call the function on those specific data partitions. It is a generalized function.

File: task_1/plotting_utils/plot_distributions.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import os
```
[the remark this answers is not in the document]
Comment (Sum02dean) [reply task_1/plotting_utils/plot_distributions.py:1]: The notebook is a scratch pad that I used for writing the function. The intention is to have all finalised plotting utilities in a single place (plotting_utils/). Not split out across  notebooks. 

I think it is better to remove it from the notebook and instead import the function from plotting_utils.py

[the remark this answers is not in the document]
Comment (Sum02dean) [reply task_1/plotting_utils/plot_distributions.py:1]: The function takes in any two samples (is general). It can be used for any data partition not just GEX. I just provided an example.

**Comment by Sum02dean** [comment]
I think we should reject this PR.
I need to decide on if I want the func in a NB or to put it in a .py file. In the latter case, I need to add init python files for loading modules, and remove the func body from NB.

I would also like to add the other stratifications that Liine mentioned to include choromosomal, cell-line and distributions and global distributions. Then increase the code within plotting_utils dir.

ToDo:
* Remove redundant code from NB
* Add module imports for plotting_utils
* Etend anlaysis to other data stratifications 
* Increase generalised code coverage of plotting_utils

These distribution analyses will help inform on the Train-Val splits and help understand the data more.

**Comment by LiineKasak** [comment]
ah okay. what disturbs me is that in my eyes the train-valid split they made is a random one in the sense that doesn't matter by which chromosomes you split it.. if you want a general comparison across labels without splitting chromosomes then you should randomly split labels after shuffling train and valid data. right now the differences might be among the chromosomes they decided to split the train and valid, not among random GEX values.

And in the end the validation set they chose probably isn't going to be ours anyways, as a k fold crossvalidation is necessary anyways imo.

No need to reject this PR just improve it to cover more cases. I propose:

- keep the util function in py but no need for commented out code. If there's an example of usage, it can be in the `if __name__ == '__main__'` block of the same file, and file paths should always be relative to the file, not absolute.
- in the notebook import the function and make several plots:

1. totally random division of labels, not by any predefined splits. so instead of splitting train and valid which is split by chr, concat the two sets, and then make a random division I guess?
2. different splits by chromosomes
3. split by cell line

And maybe these will help us decide if stratified sampling should be used somehow, but the k fold validation should be split by chromosomes anyways as said by multiple papers.

what did you mean by init files for module loading? why is there a need for them when can just import from the util file?

File: task_1/plotting_utils/plot_distributions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LiineKasak) [reply task_1/plotting_utils/plot_distributions.py:1]: > The function takes in any two samples (is general). It can be used for any data partition not just GEX. I just provided an example.

yeah I didn't look too much into it I guess, but a few more plots would be nice to get a better view of the dataset :)

File: task_1/plotting_utils/plot_distributions.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    if balance:
+        try:
+            # Check length of x1 <= x2
+            assert(x1.shape[0] <= x2.shape[0])
+
+            # Balance data
+            n_samples = np.shape(x1)[0]
+            x2 = x2.sample(n_samples)
+
+        except AssertionError as e:
```
Comment (LiineKasak) [inline task_1/plotting_utils/plot_distributions.py:34]: would be nicer if we don't impose this restriction of size for given sets, but just act accordingly to which is larger

**Comment by Sum02dean** [comment]
> what did you mean by init files for module loading? why is there a need for them when can just import from the util file?

I wanted to import a function from a different directory. ButI realised that I don't need to add  `__init__.py` I can just add the directory to path at top of NB file with:
```py
sys.path.append('../')
```
Then call 

```py
from plotting_utils.plot_distributions import plot_distributions
```
